src/hotspot/share/opto/vectorization.cpp

```diff
@@ -89,29 +87,81 @@ const char* VLoop::check_preconditions_helper() {
       _lpt->dump_head();
     }
 #endif
-    return VLoop::FAILURE_CONTROL_FLOW;
+    return VStatus::make_failure(VLoop::FAILURE_CONTROL_FLOW);
   }
 
   // Make sure the are no extra control users of the loop backedge
   if (_cl->back_control()->outcnt() != 1) {
-    return VLoop::FAILURE_BACKEDGE;
+    return VStatus::make_failure(VLoop::FAILURE_BACKEDGE);
   }
 
   // To align vector memory accesses in the main-loop, we will have to adjust
   // the pre-loop limit.
   if (_cl->is_main_loop()) {
     CountedLoopEndNode* pre_end = _cl->find_pre_loop_end();
     if (pre_end == nullptr) {
-      return VLoop::FAILURE_PRE_LOOP_LIMIT;
+      return VStatus::make_failure(VLoop::FAILURE_PRE_LOOP_LIMIT);
     }
     Node* pre_opaq1 = pre_end->limit();
     if (pre_opaq1->Opcode() != Op_Opaque1) {
-      return VLoop::FAILURE_PRE_LOOP_LIMIT;
+      return VStatus::make_failure(VLoop::FAILURE_PRE_LOOP_LIMIT);
     }
     _pre_loop_end = pre_end;
   }
 
-  return VLoop::SUCCESS;
+  return VStatus::make_success();
+}
+
+// Return true iff all submodules are loaded successfully
+bool VLoopAnalyzer::setup_submodules() {
+#ifndef PRODUCT
+  if (_vloop.is_trace_loop_analyzer()) {
+    tty->print_cr("\nVLoopAnalyzer::setup_submodules");
+    _vloop.lpt()->dump_head();
+    _vloop.cl()->dump();
+  }
+#endif
+
+  VStatus status = setup_submodules_helper();
+  if (!status.is_success()) {
+#ifndef PRODUCT
+    if (_vloop.is_trace_loop_analyzer()) {
+      tty->print_cr("\nVLoopAnalyze::setup_submodules: failed: %s", status.failure_reason());
+    }
+#endif
+    return false; // failed
+  }
+  return true; // success
+}
+
+VStatus VLoopAnalyzer::setup_submodules_helper() {
+  // Skip any loop that has not been assigned max unroll by analysis.
+  if (SuperWordLoopUnrollAnalysis && _vloop.cl()->slp_max_unroll() == 0) {
+    return VStatus::make_failure(VLoopAnalyzer::FAILURE_NO_MAX_UNROLL);
+  }
+
+  if (SuperWordReductions) {
+    _reductions.mark_reductions();
+  }
+
+  _memory_slices.find_memory_slices();
+
+  // If there is no memory slice detected, it means there is no store.
+  // If there is no reduction and no store, then we give up, because
+  // vectorization is not possible anyway (given current limitations).
+  if (!reductions().is_marked_reduction_loop() &&
```

Comment:
Since you do `_reductions.mark_reductions()` above, I suggest to also call this on `_reductions` to make it clear that it's the same thing within the context of this method.
```suggestion
  if (!_reductions.is_marked_reduction_loop() &&
```